jsdom-monkeypatch.js

```diff
@@ -0,0 +1,308 @@
+      return respecUrl;
+    }
+    else if (/\.[^\/\.]+$/.test(url.path) &&
+        !url.path.endsWith('.js') &&
+        !url.path.endsWith('.json')) {
+      //console.log(`fetch not needed for ${url.href} (not a JS/JSON file)`);
+      return null;
+    }
+    else if ((url.pathname === '/webcomponents/assets/scripts/autolink.js') ||
+        (url.href.startsWith(referrer) &&
+          !(/annotate_spec/i.test(url.pathname)) &&
+          !(/expanders/i.test(url.pathname)) &&
+          !(/bug-assist/i.test(url.pathname)) &&
+          !(/dfn/i.test(url.pathname)) &&
+          !(/section-links/i.test(url.pathname)) &&
+          !(/^\/webidl\//i.test(url.pathname)))) {
+      //console.log(`fetch useful script at ${url.href}`);
+      return url.href;
+    }
+    //console.log(`fetch not needed for ${url.href}`);
+    return null;
+  }
+
+  let urlToFetch = getUrlToFetch();
+  if (!urlToFetch) {
+    return callback(null, '');
+  }
+  fetch(urlToFetch, options)
+    .then(response => response.text())
+    .then(data => {
+      if (urlToFetch !== respecUrl) {
+          return data;
+      }
+
+      ////////////////////////////////////////////////////////////
+      // REALLY UGLY CODE WARNING
+      //
+      // Tweak Respec built code so that it can run in JSDOM.
+      //
+      // NB: Some of these lines will just break if Respec build
+      // produces slightly different code, e.g. if variables do
+      // not end up with the same name!
+      ////////////////////////////////////////////////////////////
+
+      // Remove core/highlight module because JSDOM does not yet
+      // support URL.createObjectURL
+      // https://github.com/jsdom/jsdom/issues/1721
+      // Remove core/list-sorter module because JSDOM does not yet
+      // support document.createRange
+      // https://github.com/jsdom/jsdom/blob/master/lib/jsdom/living/nodes/Document.webidl#L39
+      ["core/highlight", "core/list-sorter"]
+        .forEach(module => data = data.replace(
+          new RegExp('(define\\(\\s*"profile-w3c-common"\\s*,\\s*\\[[^\\]]+),\\s*"' + module + '"'),
+          '$1'));
+
+      // JSDOM's CSS parser does not quite like uncommon "@" rules
+      // so let's pretend they are just @media rules
+      // https://github.com/jsdom/jsdom/issues/2026
+      // (NB: this replacement is just for convenience, to avoid JSDOM reporting
+      // lengthy errors (including a full dump of the CSS) to stderr
+      data = data.replace(/@keyframes \S+? {/, '@media all {');
+      data = data.replace(/@supports \(.+?\) {/, '@media all {');
+
+      // Respec drops blank lines in Markdown, but marked.js actually
+      // needs them around <pre> tags, otherwise it produces really weird
+      // HTML (with <pre> and <p> intertwined). For some reason, this does
+      // not bother regular browsers. It does bother JSDOM though.
+      data = data.replace(/r\.createTextNode\("\\n"\)/, 'r.createTextNode("\\n\\n")');
```

Comment:
@tidoust We would be interested in fixing this in jsdom (or potentially in [parse5](https://github.com/inikulin/parse5)). Do you have a sample of the HTML which behaves differently between jsdom and other browsers?

Your list in https://github.com/w3c/respec/issues/1469#issuecomment-388560878 is very nice as well. It'd be great if you could file new issues if you find something not covered by the existing ones.

Reply:
Thanks @Zirro! I don't think there is any bug in jsdom, actually. What I get from jsdom is pretty much what I expect to get (wrong markup in, wrong markup out). The output I do not understand is the one in regular browsers... but I suspect that the wrong markup gets fixed by chance by ReSpec afterwards when it runs the `core/highlight` module (module which cannot run in jsdom). I ran out of time on this issue, so did not investigate further, but the real bug is in marked.js who produces wrong markup, not in jsdom.

I didn't file any issue in jsdom from the list because all of the current missing features either appear as real issues already, or are explicit pointed out as missing in code comments or commit messages. So that all seemed well-known already. Reffy now runs ReSpec in JSDOM (with the monkey-patching above), so we'll keep an eye on potential new problems that we may find!